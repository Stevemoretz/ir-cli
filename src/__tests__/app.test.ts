import {
    buildNormalIndex,
    buildPositionalIndex,
    combinedNormalIndex,
    getTokens,
    positionalIndexRanking,
    processString,
    queryWithNormalIndex,
    queryWithNormalIndexViaSkipPointers,
    queryWithPositionalIndex,
} from "../utils";
import merge from "deepmerge";

const demo1 =
    "On his return, Caesar married Pompeia, a granddaughter of Sulla. Their marriage ended in scandal.";
const demo2 = "It was rumored that Caesar was the father of Servilia's son";
const demo3 = "son and father relationship that was an open secret at Rome.";

describe("Boolean Queries", () => {
    describe("index", () => {
        it("posting list tokens", () => {
            expect(getTokens(1, demo1)).toMatchSnapshot();
        });
        it("posting list combinedNormalIndex", () => {
            expect(
                combinedNormalIndex([demo1, demo2, demo3])
            ).toMatchSnapshot();
        });
        it("posting list index", () => {
            expect(buildNormalIndex([demo1, demo2, demo3])).toMatchSnapshot();
        });
    });

    describe("posting list query test", () => {
        it("posting list test and of two words", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString("caesar and father");
            expect(queryWithNormalIndex(query, index)).toEqual([2]);
        });
        it("posting list test and of non existence word", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString("caesar and not_in_the_docs");
            expect(queryWithNormalIndex(query, index)).toEqual([]);
        });
        it("posting list test and of three words no match", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString("caesar and father and granddaughter");
            expect(queryWithNormalIndex(query, index)).toEqual([]);
        });
        it("posting list test and of three words one match", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString("caesar and father and it");
            expect(queryWithNormalIndex(query, index)).toEqual([2]);
        });
        it("posting list test and of three words no match but longer", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString("caesar and father and it and 6546546");
            expect(queryWithNormalIndex(query, index)).toEqual([]);
        });
        it("posting list test and of three words one match in many docs", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString(
                "caesar and father and it and rumor and s and son and servilia"
            );
            expect(queryWithNormalIndex(query, index)).toEqual([2]);
        });
    });

    describe("posting list query test with skip pointers", () => {
        it("posting list test and of two words", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString("caesar and father");
            expect(queryWithNormalIndexViaSkipPointers(query, index)).toEqual([
                2,
            ]);
        });
        it("posting list test and of non existence word", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString("caesar and not_in_the_docs");
            expect(queryWithNormalIndexViaSkipPointers(query, index)).toEqual(
                []
            );
        });
        it("posting list test and of three words no match", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString("caesar and father and granddaughter");
            expect(queryWithNormalIndexViaSkipPointers(query, index)).toEqual(
                []
            );
        });
        it("posting list test and of three words one match", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString("caesar and father and it");
            expect(queryWithNormalIndexViaSkipPointers(query, index)).toEqual([
                2,
            ]);
        });
        it("posting list test and of three words no match but longer", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString("caesar and father and it and 6546546");
            expect(queryWithNormalIndexViaSkipPointers(query, index)).toEqual(
                []
            );
        });
        it("posting list test and of three words one match in many docs", () => {
            const index = buildNormalIndex([demo1, demo2, demo3]);
            const query = processString(
                "caesar and father and it and rumor and s and son and servilia"
            );
            expect(queryWithNormalIndexViaSkipPointers(query, index)).toEqual([
                2,
            ]);
        });
    });
});

const demoPositional1 = "Caesar at the wedding";
const demoPositional2 =
    "<header>And we had Caesar at the wedding</header> Caesar at";
// const demoPositional2 = "And we had Caesar at the wedding Caesar at";
const demoPositional3 = "asdds";

describe("Proximity Queries", () => {
    describe("index", () => {
        it("index", () => {
            const index = buildPositionalIndex([
                demoPositional1,
                demoPositional2,
                demoPositional3,
            ]);
            expect(index).toMatchSnapshot();
        });
    });
    describe("posting list positional query test", () => {
        it("posting list test and of two words", () => {
            const index = buildPositionalIndex([
                demoPositional1,
                demoPositional2,
                demoPositional3,
            ]);
            const query = processString("Caesar at");
            expect(Object.keys(queryWithPositionalIndex(query, index))).toEqual(
                ["1", "2"]
            );
        });
    });
    describe("posting list positional query test ranking", () => {
        it("posting list test and of two words", () => {
            const index = buildPositionalIndex([
                demoPositional1,
                demoPositional2,
                demoPositional3,
            ]);
            const query = processString("Caesar at");
            const weightLessIndex = Object.keys(index).reduce(
                (prev, current) => {
                    const strings = current.split(".");
                    const term = strings[strings.length - 1];
                    return {
                        ...prev,
                        [term]: merge(prev[term], index[current]),
                    };
                },
                {}
            );
            const result = queryWithPositionalIndex(
                [...query],
                weightLessIndex
            );

            const ranking = positionalIndexRanking(
                result,
                index,
                weightLessIndex
            );
            expect(ranking.map((item) => item.docId)).toEqual(["2", "1"]);
        });
    });
});
