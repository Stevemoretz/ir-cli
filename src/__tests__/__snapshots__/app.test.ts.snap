// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boolean Queries index posting list combinedNormalIndex 1`] = `
Array [
  Object {
    "docId": 3,
    "index": 5,
    "term": "at",
  },
  Object {
    "docId": 1,
    "index": 2,
    "term": "caesar",
  },
  Object {
    "docId": 2,
    "index": 2,
    "term": "caesar",
  },
  Object {
    "docId": 1,
    "index": 9,
    "term": "end",
  },
  Object {
    "docId": 2,
    "index": 3,
    "term": "father",
  },
  Object {
    "docId": 3,
    "index": 1,
    "term": "father",
  },
  Object {
    "docId": 1,
    "index": 5,
    "term": "granddaughter",
  },
  Object {
    "docId": 2,
    "index": 0,
    "term": "it",
  },
  Object {
    "docId": 1,
    "index": 8,
    "term": "marriage",
  },
  Object {
    "docId": 1,
    "index": 3,
    "term": "marry",
  },
  Object {
    "docId": 1,
    "index": 0,
    "term": "on",
  },
  Object {
    "docId": 3,
    "index": 3,
    "term": "open",
  },
  Object {
    "docId": 1,
    "index": 4,
    "term": "pompeia",
  },
  Object {
    "docId": 3,
    "index": 2,
    "term": "relationship",
  },
  Object {
    "docId": 3,
    "index": 6,
    "term": "rome",
  },
  Object {
    "docId": 2,
    "index": 1,
    "term": "rumor",
  },
  Object {
    "docId": 2,
    "index": 5,
    "term": "s",
  },
  Object {
    "docId": 1,
    "index": 10,
    "term": "scandal",
  },
  Object {
    "docId": 3,
    "index": 4,
    "term": "secret",
  },
  Object {
    "docId": 2,
    "index": 4,
    "term": "servilia",
  },
  Object {
    "docId": 2,
    "index": 6,
    "term": "son",
  },
  Object {
    "docId": 3,
    "index": 0,
    "term": "son",
  },
  Object {
    "docId": 1,
    "index": 6,
    "term": "sulla",
  },
  Object {
    "docId": 1,
    "index": 7,
    "term": "their",
  },
  Object {
    "docId": 1,
    "index": 1,
    "term": "turn",
  },
]
`;

exports[`Boolean Queries index posting list index 1`] = `
Object {
  "at": Array [
    3,
  ],
  "caesar": Array [
    1,
    2,
  ],
  "end": Array [
    1,
  ],
  "father": Array [
    2,
    3,
  ],
  "granddaughter": Array [
    1,
  ],
  "it": Array [
    2,
  ],
  "marriage": Array [
    1,
  ],
  "marry": Array [
    1,
  ],
  "on": Array [
    1,
  ],
  "open": Array [
    3,
  ],
  "pompeia": Array [
    1,
  ],
  "relationship": Array [
    3,
  ],
  "rome": Array [
    3,
  ],
  "rumor": Array [
    2,
  ],
  "s": Array [
    2,
  ],
  "scandal": Array [
    1,
  ],
  "secret": Array [
    3,
  ],
  "servilia": Array [
    2,
  ],
  "son": Array [
    2,
    3,
  ],
  "sulla": Array [
    1,
  ],
  "their": Array [
    1,
  ],
  "turn": Array [
    1,
  ],
}
`;

exports[`Boolean Queries index posting list tokens 1`] = `
Array [
  Object {
    "docId": 1,
    "index": 0,
    "term": "on",
  },
  Object {
    "docId": 1,
    "index": 1,
    "term": "turn",
  },
  Object {
    "docId": 1,
    "index": 2,
    "term": "caesar",
  },
  Object {
    "docId": 1,
    "index": 3,
    "term": "marry",
  },
  Object {
    "docId": 1,
    "index": 4,
    "term": "pompeia",
  },
  Object {
    "docId": 1,
    "index": 5,
    "term": "granddaughter",
  },
  Object {
    "docId": 1,
    "index": 6,
    "term": "sulla",
  },
  Object {
    "docId": 1,
    "index": 7,
    "term": "their",
  },
  Object {
    "docId": 1,
    "index": 8,
    "term": "marriage",
  },
  Object {
    "docId": 1,
    "index": 9,
    "term": "end",
  },
  Object {
    "docId": 1,
    "index": 10,
    "term": "scandal",
  },
]
`;

exports[`Proximity Queries index index 1`] = `
Object {
  "asdds": Object {
    "3": Array [
      0,
    ],
  },
  "at": Object {
    "1": Array [
      1,
    ],
    "2": Array [
      7,
    ],
  },
  "caesar": Object {
    "1": Array [
      0,
    ],
    "2": Array [
      6,
    ],
  },
  "header.and": Object {
    "2": Array [
      0,
    ],
  },
  "header.at": Object {
    "2": Array [
      4,
    ],
  },
  "header.caesar": Object {
    "2": Array [
      3,
    ],
  },
  "header.have": Object {
    "2": Array [
      2,
    ],
  },
  "header.we": Object {
    "2": Array [
      1,
    ],
  },
  "header.wed": Object {
    "2": Array [
      5,
    ],
  },
  "wed": Object {
    "1": Array [
      2,
    ],
  },
}
`;
